### PR TITLE
feat: update swapper hooks to use shared approval hooks/queries/mutations

### LINKED ISSUE #7427: USDT reset gas state bug
### Overview

It costs more gas to do an approval when the current approval is `0` than when it is non-zero.

This is causing a gas estimation bug due to a state change on chain occurring after the initial gas estimate is done.

### References and additional details

A Jam of max reproducing: https://jam.dev/c/4597be3b-1990-43e1-a3ea-b6860bb35cd6

### Acceptance Criteria

When doing a USDT trade that requires an approval reset, we should properly re-calculate the gas cost of the approval _after_ the reset has been done (instead of using the value on initial render, which is too low because the state on chain will have changed by the time it's executed).

Once this bug is resolved, the closing PR should also enable the `REACT_APP_FEATURE_USDT_APPROVAL_RESET` feature flag.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by kaladinlight**
## Description

There has been some work to add common approval logic recently that should be leveraged in swapper. Update to use and clean up all of the things.

- Fixes the active step indicator icon status
- Fees are estimated JIT before broadcast so there are no stale fee issues anymore.

~~Leaving in draft state for now as to not conflict with https://github.com/shapeshift/web/pull/7237~~

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/7427

High - total rework of approval logic in swapper and slight cleanup to common approval logic.

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

- Scrutinize approval flows for swapper (standard approval and allowance reset via USDT)
- Spot check approval flows elsewhere (rFox, Thorchain LP)

### Engineering

:point_up:

### Operations

:point_up:

I did not test multihop if you would be able to give multihop with approval flows a run through :pray: 

## Screenshots (if applicable)

https://github.com/user-attachments/assets/9d4c8cdc-72fd-4908-b430-38dcec43e0aa

https://github.com/user-attachments/assets/1ee09d4a-d178-4a5d-a547-7474b5ad5d06

https://github.com/user-attachments/assets/75132aa3-8674-4b65-9b39-c72824f436c7

### CODE REVIEW

**Comment by kaladinlight** [comment]
Check to see if https://github.com/shapeshift/web/issues/7427 is resolved with this update

**Comment by kaladinlight** [comment]
> Check to see if #7427 is resolved with this update

This should be fixed now. We are not using stale fee estimates from display for broadcast, instead estimating JIT before broadcast for most accurate fees possible.

**Review by 0xApotheosis — APPROVED** [review]
✅ **Exact approval flow**

https://snowtrace.dev/tx/0x6e5ee31ccb57e4e6908541effea73b98737cc1c47b8f3f69310a6ce36b9f9a0c
https://snowtrace.dev/tx/0x1b934532565074ecc3c910ff1d1dac2e1c0155111232470e3887419a8365ec66

Screenshot 2024-07-31 at 3 52 49 PM

✅ **Unlimited approval flow**

https://snowtrace.dev/tx/0xd5dc0e38205ec647f528a87d66dd787e70161b52a6a96e4fe8b1a0719c337c52
https://snowtrace.dev/tx/0x4851271f82453538fbd24ff00f7e4310f28d274e665aa31231f657eba3ac408f

Screenshot 2024-07-31 at 3 55 07 PM

✅ **No approval needed**

https://snowtrace.dev/tx/0x8ea1d250553d4cc9b591f3e76061fdc2168651cc8dcc9c936ded728a81df2e41

Screenshot 2024-07-31 at 3 56 03 PM

✅ **USDT reset required, flag off**

No approval reset step shown, no delta from production - feature flag works as expected

Screenshot 2024-07-31 at 4 02 56 PM

✅ **USDT reset required, flag on**

https://snowtrace.dev/tx/0x3ac1bdd5cdf099f8f5b70c0ee5f679df382e5e5458edaae7719742037a2e8f30
https://snowtrace.dev/tx/0xf992296e945a2efe0fe66de6a3265adbdd13dad33b980846d010e1334da0a45a
https://snowtrace.dev/tx/0x4f70578292e391f7790105455ce289dc959a8a1ac84e2ed7ff4ae4bfa137562b

Screenshot 2024-07-31 at 4 08 07 PM

✅ Other assets are unaffected with the USDT flag on

Screenshot 2024-07-31 at 4 08 59 PM

File: src/lib/utils/evm/approve.ts [hunk 1]
```diff
@@ -1,46 +1,40 @@
...
-    supportsEIP1559: supportsETH(wallet) && (await wallet.ethSupportsEIP1559()),
+  const data = getApproveContractData({
+    approvalAmountCryptoBaseUnit: amountCryptoBaseUnit,
+    spender,
+    to,
+    chainId,
   })
 
-  const buildCustomTxInput = {
+  const buildCustomTxInput = await createBuildCustomTxInput({
```
Comment (gomesalexandre) [inline src/lib/utils/evm/approve.ts:30]: 💜

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useIsApprovalInitiallyNeeded.tsx [hunk 2]
```diff
@@ -114,7 +112,7 @@ export const useIsApprovalInitiallyNeeded = () => {
   ])
 
   const result = useMemo(
-    () => ({ isLoading: isFirstHopLoading || (isMultiHopTrade && isSecondHopLoading) }),
+    () => ({ isLoading: isFirstHopLoading || (Boolean(isMultiHopTrade) && isSecondHopLoading) }),
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useIsApprovalInitiallyNeeded.tsx:115]: cleanup: looking at `selectIsActiveQuoteMultiHop` consumptions, there doesn't seem to be any need for it to return `boolean | undefined` if we want to change that instead

**Review by gomesalexandre — COMMENTED** [review]
Conceptual pass, looks stellar! Nothing standing out, and those new utils look sweet.

Functional pass TBD in the next review.

**Review by gomesalexandre — APPROVED** [review]
Tested locally, does what it says on the box!

https://gist.github.com/gomesalexandre/3260eb640b7c4dcc8061730ecd0f4731

While smoke testing, noticed savers are borked - though not introduced by this PR. Probably worth sanity-checking USDC Txs in a follow-up so we can implement the same reset logic across domains, but as far as this PR goes... get in!

**Pull request merged**